File: src/Tests/Behavioral/PointerToPointer/PointerToPointer.go
package main

import "fmt"

type Buffer struct {
	buf      []byte
	off      int
	lastRead int8
}

const (
	opRead    int8 = -1
	opInvalid int8 = 0
)

func main() {
	var a int
	var ptr *int
	var pptr **int
	var ppptr ***int

	a = 3000

	/* take the address of var */
	ptr = &a

	/* take the address of ptr using address of operator & */
	pptr = &ptr
	ppptr = &pptr

	/* take the value using pptr */
	fmt.Printf("Value of a = %d\n", a)
	PrintValPtr(ptr)
	fmt.Printf("Main-function return value available at *ptr = %d\n", *EscapePrintValPtr(ptr))
	fmt.Printf("Main-function updated value available at *ptr = %d\n", *ptr)
	PrintValPtr2Ptr(pptr)
	PrintValPtr2Ptr2Ptr(ppptr)

	a = 1900

	fmt.Printf("Value of a = %d\n", a)
	PrintValPtr(ptr)
	fmt.Printf("Main-function return value available at *ptr = %d\n", *EscapePrintValPtr(ptr))
	fmt.Printf("Main-function updated value available at *ptr = %d\n", *ptr)
	PrintValPtr2Ptr(pptr)
	PrintValPtr2Ptr2Ptr(ppptr)

	b := Buffer{}
	PrintValPtr(&b.off)
	PrintValPtr(&b.off)

	derefElements()
}

func (b *Buffer) Read(p []byte) (n int, err error) {
	b.lastRead = opInvalid
	b.off += n
	if n > 0 {
		b.lastRead = opRead
	}

	(&Buffer{buf: p}).Read(p)

	return n, nil
}

func NewBuffer(buf []byte) (b1 *Buffer) {
	return &Buffer{buf: buf}
}

func PrintValPtr(ptr *int) {
	fmt.Printf("Value available at *ptr = %d\n", *ptr)
	*ptr++
}

func EscapePrintValPtr(out *int) *int {
	fmt.Printf("Value available at *ptr = %d\n", *out)
	i := 99
	out = &i
	fmt.Printf("Intra-function updated value available at *ptr = %d\n", *out)
	PrintValPtr(out)
	return out
}

func PrintValPtr2Ptr(pptr **int) {
	fmt.Printf("Value available at **pptr = %d\n", **pptr)
}

func PrintValPtr2Ptr2Ptr(ppptr ***int) {
	fmt.Printf("Value available at ***pptr = %d\n", ***ppptr)
}

// derefElements: a deref of a SLICE-OF-POINTERS element (`*temps[depth]`, math/big
// natdiv): the param-deref shortcut must not elide the element's own dereference (the
// raw ж<T> leaked into every use, CS1929/CS1503 ×29). Write through and read back.
func derefElements() {
	a1, a2 := 10, 20
	boxes := []*int{&a1, &a2}
	*boxes[0] = *boxes[0] + 5
	*boxes[1] = readBox(boxes, 1) * 3
	fmt.Println(a1, a2, *boxes[0]) // 15 60 15
}

func readBox(items []*int, i int) int {
	return *items[i]
}

File: src/Tests/Behavioral/PointerToPointer/PointerToPointer.cs
namespace go;

using fmt = fmt_package;

partial class main_package {

[GoType] partial struct Buffer {
    public slice<byte> buf;
    public nint off;
    public int8 lastRead;
}

private const int8 opRead = -1;
private const int8 opInvalid = 0;
private static void Main() {
    ref var a = ref heap(new nint(), out var Ꮡa);
    ж<nint> ptr = default!;
    ж<ж<nint>> pptr = default!;
    ж<ж<ж<nint>>> ppptr = default!;
    a = 3000;
    ptr = Ꮡa;
    pptr = Ꮡ(ptr);
    ppptr = Ꮡ(pptr);
    fmt.Printf("Value of a = %d\n"u8, a);
    PrintValPtr(ptr);
    fmt.Printf("Main-function return value available at *ptr = %d\n"u8, EscapePrintValPtr(ptr).val);
    fmt.Printf("Main-function updated value available at *ptr = %d\n"u8, ptr.val);
    PrintValPtr2Ptr(pptr);
    PrintValPtr2Ptr2Ptr(ppptr);
    a = 1900;
    fmt.Printf("Value of a = %d\n"u8, a);
    PrintValPtr(ptr);
    fmt.Printf("Main-function return value available at *ptr = %d\n"u8, EscapePrintValPtr(ptr).val);
    fmt.Printf("Main-function updated value available at *ptr = %d\n"u8, ptr.val);
    PrintValPtr2Ptr(pptr);
    PrintValPtr2Ptr2Ptr(ppptr);
    ref var b = ref heap<Buffer>(out var Ꮡb);
    b = new Buffer(nil);
    PrintValPtr(Ꮡb.of(Buffer.Ꮡoff));
    PrintValPtr(Ꮡb.of(Buffer.Ꮡoff));
}

[GoRecv] public static (nint n, error err) Read(this ref Buffer b, slice<byte> p) {
    nint n = default!;
    error err = default!;

    b.lastRead = opInvalid;
    b.off += n;
    if (n > 0) {
        b.lastRead = opRead;
    }
    (Ꮡ(new Buffer(buf: p))).Read(p);
    return (n, default!);
}

public static ж<Buffer> /*b1*/ NewBuffer(slice<byte> buf) {
    ж<Buffer> b1 = default!;

    return Ꮡ(new Buffer(buf: buf));
}

public static void PrintValPtr(ж<nint> Ꮡptr) {
    ref var ptr = ref Ꮡptr.val;

    fmt.Printf("Value available at *ptr = %d\n"u8, ptr);
    ptr++;
}

public static ж<nint> EscapePrintValPtr(ж<nint> Ꮡout) {
    ref var @out = ref Ꮡout.val;

    fmt.Printf("Value available at *ptr = %d\n"u8, @out);
    ref var i = ref heap<nint>(out var Ꮡi);
    i = 99;
    Ꮡout = Ꮡi; @out = ref Ꮡout.val;
    fmt.Printf("Intra-function updated value available at *ptr = %d\n"u8, @out);
    PrintValPtr(Ꮡout);
    return Ꮡout;
}

public static void PrintValPtr2Ptr(ж<ж<nint>> Ꮡpptr) {
    ref var pptr = ref Ꮡpptr.val;

    fmt.Printf("Value available at **pptr = %d\n"u8, pptr.val);
}

public static void PrintValPtr2Ptr2Ptr(ж<ж<ж<nint>>> Ꮡppptr) {
    ref var ppptr = ref Ꮡppptr.val;

    fmt.Printf("Value available at ***pptr = %d\n"u8, ppptr.val.val);
}

} // end main_package
